Objects/dictobject.c

```diff
@@ -6852,46 +6832,55 @@ PyObject_VisitManagedDict(PyObject *obj, visitproc visit, void *arg)
     if((tp->tp_flags & Py_TPFLAGS_MANAGED_DICT) == 0) {
         return 0;
     }
-    assert(tp->tp_dictoffset);
-    PyDictOrValues dorv = *_PyObject_DictOrValuesPointer(obj);
-    if (_PyDictOrValues_IsValues(dorv)) {
-        PyDictValues *values = _PyDictOrValues_GetValues(dorv);
-        PyDictKeysObject *keys = CACHED_KEYS(tp);
-        for (Py_ssize_t i = 0; i < keys->dk_nentries; i++) {
-            Py_VISIT(values->values[i]);
+    if (tp->tp_flags & Py_TPFLAGS_INLINE_VALUES) {
+        PyDictValues *values = _PyObject_InlineValues(obj);
+        if (values->valid) {
+            for (Py_ssize_t i = 0; i < values->capacity; i++) {
+                Py_VISIT(values->values[i]);
+            }
+            return 0;
         }
     }
-    else {
-        PyObject *dict = _PyDictOrValues_GetDict(dorv);
-        Py_VISIT(dict);
-    }
+    Py_VISIT(_PyObject_ManagedDictPointer(obj)->dict);
     return 0;
 }
 
 void
 PyObject_ClearManagedDict(PyObject *obj)
 {
+    assert(_PyObject_InlineValuesConsistencyCheck(obj));
     PyTypeObject *tp = Py_TYPE(obj);
-    if((tp->tp_flags & Py_TPFLAGS_MANAGED_DICT) == 0) {
-        return;
-    }
-    PyDictOrValues *dorv_ptr = _PyObject_DictOrValuesPointer(obj);
-    if (_PyDictOrValues_IsValues(*dorv_ptr)) {
-        PyDictValues *values = _PyDictOrValues_GetValues(*dorv_ptr);
-        PyDictKeysObject *keys = CACHED_KEYS(tp);
-        for (Py_ssize_t i = 0; i < keys->dk_nentries; i++) {
-            Py_CLEAR(values->values[i]);
+    Py_CLEAR(_PyObject_ManagedDictPointer(obj)->dict);
+    if (tp->tp_flags & Py_TPFLAGS_INLINE_VALUES) {
```

Comment:
This can leave the values in an invalid state if it's called from `tp_clear`. There may still be a dict referring to the inline values. That's causing the crash in `test_asyncio`.

This can happen in both the default and free-threaded build, but it depends on which object gets `tp_clear`'d first.

Reply:
@DinoV points out that the implementation earlier in the PR before https://github.com/python/cpython/pull/116115/commits/895a94460a1e9cb1f6e29a8453fc5b2c4e700ddd was correct.

Reply:
In theory, the only dict that can be pointing to the inline values is the one pointed to by `_PyObject_ManagedDictPointer(obj)->dict`.

I think the bug found by `test_asyncio` was that we weren't properly detaching the dict before setting a new one in `obj.__dict__ = ...`.